### ISSUE TITLE
 Initial design picker screen: add categories 

### ISSUE THREAD

**Issue by enejb** [wordpress-mobile/gutenberg-mobile#3125]
### Why are we doing this?

We want to make it easier for our users to find the design that they are looking for by presenting them in a categorized list. Since we are planning to add a lot more designs in the future.

### Goals

- Update the current design picker to behave in the same way as the current categorized page picker (Starter Page Template Picker). 
- The user should be able to select a category. Once selected only the selected categories design should be available to be chosen.
- Once a design is selected the user should be presented with the previous “Preview” and “Choose” buttons. 
- If the user selects preview they should be able to preview the designs. The same way they were able to before.
- If the user selects the  “Choose” button they should be presented with the choose the domain screen. (The same as before)

### Changes that will need to happen to the apps.

1. Update the copy of the description on the screen from "Pick your favourite homepage layout. You can edit and customize it later." to
"Get started with a pre-made layout and customize it to fit your needs"
2. The new categorized design picker will be very similar to the Page picker design. 
3. Our new layouts will be grouped horizontally in the same way as the Starter Page Picker.
4. Note that the aspect ratio of the preview image is staying the same. 

#### Endpoint Changes
We are also implementing a new endpoint. The endpoint is being implemented in D56541-code.

The endpoint will be very similar to the common-block-layouts and will return a list of design + categories. The URL that we are planning to use will be. We have decided that for Android we will return the new "Single page designs + the original designs. While for iOS we will just return the original designs for now. 
https://public-api.wordpress.com/wpcom/v2/common-starter-site-designs/

- Query Parameters:
    - `type=mobile` - to signify this is the mobile apps
    - `group` - You will be able to pass in a group through as a query parameter in order to return the correct group of layouts.
        - By default the endpoint will ____?____ 
        - **iOS** should set this to equal `original` to exclude the old layouts.
        - **Android** should set this to equal `original,single-page`
    - `scale` - This works the same as it does for the common-block-layouts
    - `preview_width` - This works the same as it does for the common-block-layouts

<details>
<summary>Screenshot of the new designs.</summary>

![image](https://user-images.githubusercontent.com/115071/107560737-968aad00-6b92-11eb-96da-f2437a7f573d.png)

</details>

### Additional context 

This Issue is part of the single page layout project. pb3aDo-IP-p2
Internal proposal and discussion.  pb3aDo-Lz-p2

**Comment by chipsnyder** [comment]
@antonis @mkevins 

> - iOS should set this to equal `original` to exclude the old layouts.
> - Android should set this to equal `original,single-page`

After chatting with Eli she suggested switching this requirement so that iOS gets the new `single-page` layouts and Android has the old. The thinking is by having it on iOS we'll be able to get more internal beta testers. Can we make that happen?

Follow up on that ^ 
@enejb what will be the behavior if we don't pass a value for `group`? I see I left a question mark in the description. I'm thinking maybe we have that act as `original` then leave Android blank. If we decide to roll out the layouts to both platforms then we can then update "blank" to be `original,single-page` WDYT?

**Comment by enejb** [comment]
I think we should be more intentional with the groups that we pass in, maybe having groups that describe what 
Maybe instead of passing in groups that describe the content. We should pass in something else? 

Currently we still have some designs that are there to support versions of the apps that use the "blog, business, portfolio, professional" selections even though we don't use the designs anymore. Currently when you do not pass in a group you get all the designs.

In short I think we should pass in different groups but maybe call them something else so that they don't necessary reflect the content but more the intent better. Any suggestions here?

**Comment by chipsnyder** [comment]
I'm thinking of the `group` variable as more of an A/B test scenario for instances where we decide we want to check the effectiveness of a new set of layouts. 

So no parameter would be the control, and `A` then with a value would be `B` if we decide that `B` was effective, we would want to roll it out to all users like we would for an A/B test. 

This way, we can add new designs without making a code change or A/B test ones if we feel there is a need. 

>Currently we still have some designs that are there to support versions of the apps that use the "blog, business, portfolio, professional" selections even though we don't use the designs anymore. Currently when you do not pass in a group you get all the designs.

So I would suggest having no parameter act as `original` and then if we decide `single-page` is doing well we can roll it out to all with an API change only. 

>I think we should be more intentional with the groups that we pass in

Just so I can understand is there a potential problem or goal that you are trying to solve with this approach?

**Comment by enejb** [comment]
> Just so I can understand is there a potential problem or goal that you are trying to solve with this approach?

There are a couple of problems that I am trying to solve. 
1. Make the api return the expected results. 
2. Make it easier to make updates in the future. 

One issue that I want to avoid is that return designs that might not be compatible with the version of the app but still have things be generic enough that we don't have to manage the designs for each version. 

In the future as editor will have more blocks supported we might want to ship layouts to just the versions of the app that have the particular block support. 

If we don't have meaningful groups in the api request then that becomes challenging. 
For example in future we could add a new few new designs and some of them might be using a more new block. If we have the older version of the app requesting the "original" group we could make sure that the new designs with the new blocks don't end up in those apps my not adding them to the group. 

The current case when you don't have the group specified returns would always return all the designs. If we limit the original Which would not be the case any more if you want to not ship a specific layout to older app versions. 

I think in the end its all very similar and we can go either way and still make it work. I just think if we can provide a bit more info to the endpoint then we can make it easier to guard things in the future iterations.

**Comment by chipsnyder** [comment]
Thank you for the explanation @enejb that helps me get into your frame of mind. I agree it sounds like we're approaching the same goals but from different angles. 

It sounds like you were thinking of treating `group` as a long-running entity that we'll continue to maintain which will be used to solve the problem of us adding templates in the future that use an unsupported block type. `jetpack/layout-grid` might be a likely candidate for this soon. 

I was thinking of treating `group` as a short-running entity that would be used for A/B testing as we determine the impact of new designs. As we determine the viability of these layouts we would deprecate the group and reroute any requests that include an old group as default. However, I wasn't considering the unsupported block type problem you bring up.

I have a few concerns about using the group to add new layouts and be the flag for supported blocks. 

**Releasing new layouts** 
Lets say we have a scenario where we want to increase the number of single-page layouts, some use `jetpack/layout-grid` but not all of the new ones, and we'd like to add them to the starter page (new page) screen as well. Our goal for this new project is to get them out as soon as possible because we're confident they'll add value. 

Following the group pattern would we:
1. Create a new group `single-page-2` and add all of the new layouts there?
1. Create a new group `single-page-2` and add the layouts that have `jetpack/layout-grid` there and the others to the other group? 

I believe the possible scenarios here increase as we consider there could be apps: 
1. That don't support `jetpack/layout-grid`  that may have the `single-page` group
2. That do support `jetpack/layout-grid` that may have the `single-page` group
     - Given the way blocks are filtered in the starter page template screen, this group would have those layouts available when they create a new page but not a new site. 
3. That do support `jetpack/layout-grid` that may have the `single-page` group plus the result of the decision above.

Would there be a reason to not roll out the new layouts to Group 2? 

**Diverging mechanisms for unsupported blocks**

On the starter page template screen we send a list of supported blocks that we use to filter templates that have unsupported blocks. I'm concerned that `group` doesn't provide a clear analogy that this is how it's being used. Have we been seeing issues with that parameter? And should we be approaching this problem more holistically and thinking of that endpoint as well?

**Comment by enejb** [comment]
Thanks for your thoughtful response. I think I understand the `group` as a short-running entry better now. 

Just to clarify 

**Releasing new layouts** 
If we go with the group scenario and we want to increase the number of single-page layouts, some use `jetpack/layout-grid` but not all of the new ones, and we'd like to add them to the starter page (new page) screen as well. Our goal for this new project is to get them out as soon as possible because we're confident they'll add value.

We would do the following:
1. Add the layouts that do not have `jetpack/layout-grid` to the `original` group. This way any app that uses that group in the api would recieve that layout. 
2. We would create a new group. for example `layout-grid` that would contain the new designs that DO have `jetpack/layout-grid` component. We will also have to update both apps to request the new group. 

Over time we could deprecate `layout-grid` group and just move the designs to the original group. We could do that with any other group that we don't consider to be "experimental" such as the current `single-page` group.

**Diverging mechanisms for unsupported blocks**
Have we been seeing issues with that parameter? 
Not that I know. I am a bit hesitant to introduce it to the new endpoint mostly because it can complicate some things. 
in the starter page template endpoint we have the content as something that we return in the endpoint. And so it is pretty trivial to loop though it and only return the pages that have do have particular blocks. Layout designs are made up of multiple pages. So this part is a bit more tricky. 

The main reason why we use the supporter blocks parameter on the page layouts endpoint is because we wanted to use layouts that are currently designed for the web. All the site layout are that we have are designed for mobile specifically. 

We can still introduce it but I think since we don't currently return any designs that have any unsupported blocks. I think it would just add complexity that we don't need :)

**Comment by chipsnyder** [comment]
Thank you for the response @enejb 

> We will also have to update both apps to request the new group.

This is the step that I was hoping to avoid by using the supported blocks parameter. Although a code change would be trivial I would like to be able to set up a system in the future that doesn't require developer input to "activate" new layouts. 

> We can still introduce it but I think since we don't currently return any designs that have any unsupported blocks. I think it would just add complexity that we don't need :)

I agree with this. Taking on the unsupported blocks problem right now is probably added complexity. My main focus was making sure we were on the same page for how to handle the scenario for when we decide we no longer need to segment `original` from `single-page`. 

It sounds like we are in agreement that once `single-page` isn't "experimental" then we can merge those layouts with `original`. 

This brings me back a bit full circle `original` is now a set of layouts that will change over time so I felt a good solution would be to have the no parameter version be the set that changes over time. `original` seems very time-specific. Maybe we could instead rename this to `mobile` and that will meet both of our concerns. WDYT?

**Comment by enejb** [comment]
I think you are right that the group query parameter is probably not the best way to capture the if a design is supported by the editor or now. 

Lets remove the 'original' part for now it for now and return the 'original' group by default.

**Comment by antonis** [comment]
Thank you @enejb and @chipsnyder for sorting this out 🙇 

> I think you are right that the group query parameter is probably not the best way to capture the if a design is supported by the editor or now.
> 
> Lets remove the 'original' part for now it for now and return the 'original' group by default.

Just to make sure that we are all in the same page:
* The iOS app will just pass the `single-page` group
* The Android app will pass no group at all

Correct? cc @mkevins

**Comment by enejb** [comment]
Yes that is correct. 

Now that the api endpoint returns all the designs when no group is passed. You will also see the single-page designs but once the fix is in place that will not be the case. 
Will try to work on the the fix for that soon.

**Comment by chipsnyder** [comment]
Awesome thank you @enejb and @antonis and thank you for your time @enejb to hash that out 🙇 


One (hopefully) small request from my testing of the iOS build.
@enejb while we're working on the fix would it be possible to group the `single-page` designs at the top of the list when they're included?

**Comment by enejb** [comment]
Yes we can definitely make that happen. :)

**Comment by enejb** [comment]
After tying to implement the changes and chatting with @mkevins . 

I think if we are passing in groups that we should list all the groups out. So instead of just passing in `single-page` we would also pass in the `original`. Which we are planning to rename to `stable` which plays well with the intend of having things be final. 

so to clarify again ( sorry )

Android: Will not pass a group at all. 
iOS: passes in `group=single-page,stable`

See for the proposed changed on the .com side of things
D57724-code

**Comment by chipsnyder** [comment]
This sounds good to me. I think `stable` is a good name as well thankyou @enejb

**Comment by enejb** [comment]
Just a quick note that D57724-code has been deployed.
